src/Geolocation/Geolocation.php

```diff
@@ -0,0 +1,336 @@
+					'userLocation' => $userLocation,
+				]);
+				return $formId;
+			}
+
+			// If we have set default locations but no match return empty form.
+			Helper::logger([
+				'geolocation' => 'Locations doesn\'t exists, default location doesn\'t match. Outputing nothing.',
+				'formIdOriginal' => $formId,
+				'formIdUsed' => '',
+				'userLocation' => $userLocation,
+			]);
+			return '';
+		}
+
+		// Final fallback if the user has no locations, no default locations or thery don't match just return the current form.
+		Helper::logger([
+			'geolocation' => 'Final fallback that returns current form. Outputing original form.',
+			'formIdOriginal' => $formId,
+			'formIdUsed' => $formId,
+			'userLocation' => $userLocation,
+		]);
+		return $formId;
+	}
+
+	/**
+	 * Get all countrie lists from the manifest.json.
+	 *
+	 * @return array<mixed>
+	 */
+	public function getCountries(): array
+	{
+		$output = [
+			[
+				'label' => __('Europe', 'eightshift-forms'),
+				'value' => 'europe',
+				'group' => [
+					'AT', 'BE', 'BG', 'HR', 'CY', 'CZ', 'DK', 'EE', 'FI', 'FR', 'DE', 'GR',
+					'HU', 'IE', 'IT', 'LV', 'LT', 'LU', 'MT', 'NL', 'PL', 'PT', 'RO', 'SK',
+					'SI', 'ES', 'SE', 'AL', 'AD', 'AM', 'BY', 'BA', 'FO', 'GE', 'GI', 'IS',
+					'IM', 'XK', 'LI', 'MK', 'MD', 'MC', 'NO', 'RU', 'SM', 'RS', 'CH', 'TR',
+					'UA', 'GB', 'VA',
+				],
+			],
+			[
+				'label' => __('European Union', 'eightshift-forms'),
+				'value' => 'european-union',
+				'group' => [
+					'BE', 'EL', 'LT', 'PT', 'BG', 'ES', 'LU', 'RO', 'CZ',
+					'FR', 'HU', 'SI', 'DK', 'HR', 'MT', 'SK', 'DE', 'IT',
+					'NL', 'FI', 'EE', 'CY', 'AT', 'SE', 'IE', 'LV', 'PL',
+				],
+			],
+			[
+				'label' => __('Ex Yugoslavia', 'eightshift-forms'),
+				'value' => 'ex-yugoslavia',
+				'group' => [
+					'HR', 'RS', 'BA', 'ME', 'SI'
+				],
+			],
+		];
+
+		// Save to internal cache so we don't read manifest all the time.
+		if (!$this->countries) {
+			$this->countries = Components::getManifest(__DIR__);
+		}
+
+		foreach ($this->countries as $country) {
+			$code = $country['Code'];
+
+			$output[] = [
+				'label' => $country['Name'],
+				'value' => $code,
+				'group' => [
+					strtoupper($code),
+				],
+			];
+		}
+
+		// Provide custom countries.
+		$filterName = Filters::getGeolocationFilterName('countries');
+		if (has_filter($filterName)) {
+			$output = apply_filters($filterName, $output ?? []);
+		}
+
+		return $output;
+	}
+
+	/**
+	 * Get Country code group
+	 *
+	 * @param string $value Code name to check.
+	 *
+	 * @return array<string>
+	 */
+	private function getCountryGroup(string $value): array
+	{
+		$country = array_filter(
+			$this->getCountries(),
+			static function ($item) use ($value) {
+				$itemValue = $item['value'] ?? '';
+				return $itemValue === $value;
+			}
+		) ?? [];
+
+		if (!$country) {
+			return [];
+		}
+
+		$country = reset($country) ?? [];
+
+		$group = $country['group'] ?? [];
+
+		return array_flip($group) ?? [];
```

Comment:
I'm wondering if early returns would be better here, instead of defaulting to an empty array all the time. Would save on the processing a bit.